### PR TITLE
[WIP] ACNN Benchmark

### PR DESCRIPTION
**Pull request by ncfrey**
I am attempting to add a benchmark for the newly fixed `AtomicConvModel` on the PDBbind dataset. I think I've implemented everything needed, but when the benchmarking tries to `load_dataset`, all the featurizations fail. I've confirmed that the corresponding MolNet loader works correctly outside of the scripts and I've tried multiple featurizers. 

@mufeili and @yuanqidu, if you get a chance to take a look at my changes, I would appreciate any guidance while I continue to work on this benchmark. And if you see any obvious errors that would be causing the featurization step to fail, please let me know!

### CODE REVIEW

**Comment by mufeili** [comment]
Did you simply run `python acnn.py -hs`? What error messages did you see?

**Comment by ncfrey** [comment]
Hi @mufeili, yes, I just encountered that error as well and pushed a change to add support for Keras models.

 Curiously, when I run `python acnn.py -hs` locally, it runs with no problem (but I'm on a cpu so it's very slow), but when I git clone my branch to Google Colab and run there, I encounter the featurization error.

**Comment by ncfrey** [comment]
This issue was resolved with `condacolab` and some manual dependency installation. I'm continuing to debug the benchmarking with `KerasModel` and `AtomicConvModel` and will hopefully have results soon.

**Comment by ncfrey** [comment]
I successfully completed a hyperparameter search and benchmark in Colab. I think this is ready for review @rbharath

File: README.md [hunk 1]
```diff
@@ -32,6 +32,12 @@
 | 1    | Random Forest | 1024-bit ECFP4 | 1.3178 +- 0.0081 | 0.6716 +- 0.0059 | [Mufei Li](mufeili1996@gmail.com) | [Paper](https://www.stat.berkeley.edu/~breiman/randomforest2001.pdf), [Code](./examples) | Dec 26th, 2020 |
 | 2    | GCN           | GraphConv      | 1.6450 +- 0.1325 | 0.5244 +- 0.0200 | [Mufei Li](mufeili1996@gmail.com) | [Paper](https://arxiv.org/abs/1609.02907), [Code](./examples)                            | Dec 26th, 2020 |
 
+### PDBbind (core)
+
+| Rank | Model         | Featurization  | Test RMSE     | Validation RMSE | Contact                           | References                                                                               | Date           |
```
Comment (rbharath) [inline README.md:37]: Could we also add R^2 here? I think RMS is more physically relevant, but R^2 is what we've used in the original ACNN paper

File: examples/acnn.py [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+  all_run_test_metrics = []
+
+  for _ in range(args['num_runs']):
+    # Model
+    model = load_model(save_path, args, tasks, hyperparams)
+    # Object for early stop tracking
+    stopper = EarlyStopper(save_path, args['metric'], args['patience'])
+
+    # 100 for maximum number of epochs
+    for epoch in range(100):
```
Comment (rbharath) [inline examples/acnn.py:67]: Should we make the number of epochs a parameter?

**Review by rbharath — COMMENTED** [review]
This is great! I've added a couple minor comments.

@mufeili It would be good to get your review here as well :)

File: examples/acnn.py [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+from hyperopt import hp, fmin, tpe
+from shutil import copyfile
+
+from utils import init_trial_path, load_dataset, EarlyStopper
+
+
+def load_model(save_path, args, tasks, hyperparams):
+  if args['dataset'] in ['PDBbind']:
+    mode = 'regression'
+    n_classes = None
```
Comment (mufeili) [inline examples/acnn.py:15]: Not used anywhere

File: examples/utils.py [hunk 4]
```diff
@@ -118,18 +137,28 @@ def __init__(self, save_path, metric, patience):
...
+      if type(model).__bases__[0] == TorchModel:
+        torch.save(model.model.state_dict(), self.save_path + '/early_stop.pt')
+      else:  # KerasModel
+        model.model.save(self.save_path + '/early_stop')
     else:
       self.patience_count += 1
 
     return self.patience_count == self.max_patience
 
   def load_state_dict(self, model):
```
Comment (mufeili) [inline examples/utils.py:160]: I wonder if it's better to use a single API for `load_state_dict` and `load_keras_model`.

File: examples/acnn.py (hunk 2 above)
Comment (ncfrey) [reply examples/acnn.py:67]: I kept this to be consistent with the GNN script, and because with early stopping the training never actually hits 100 epochs.

File: examples/utils.py (hunk 4 above)
Comment (ncfrey) [reply examples/utils.py:160]: I know that all classes that inherit from `KerasModel` have a `restore()` method and do automatic checkpointing, so I'm actually not even using the `model.model.save()` for anything here. I'm not sure if `TorchModel` has a standard way of restoring, but that could be a good thing to standardize in a future PR.

Comment (mufeili) [reply examples/utils.py:160]: Got it.

**Comment by ncfrey** [comment]
I added an R^2 option to the available metrics and repeated the benchmarking run. Interestingly, running the bayesian optimization with a different metric to optimize yielded a different set of optimal hyperparameters.
The results were also quite good! R^2 = 0.54 for test set and 0.60 for validation set.

I think this is ready for a final review @rbharath.

**Review by rbharath — APPROVED** [review]
LGTM! Feel free to merge in this PR and rebase or just merge in the larger joint PR
